radix-engine-interface/src/data/value_serializer.rs

```diff
@@ -0,0 +1,667 @@
+            serializer,
+            context,
+            "NonFungibleAddress",
+            &format!("{}", value),
+        ),
+        // Uninterpreted
+        ScryptoCustomValue::Hash(value) => {
+            serialize_type_name_tagged_value(serializer, context, "Hash", &format!("{}", value))
+        }
+        ScryptoCustomValue::EcdsaSecp256k1PublicKey(value) => serialize_type_name_tagged_value(
+            serializer,
+            context,
+            "EcdsaSecp256k1PublicKey",
+            &format!("{}", value),
+        ),
+        ScryptoCustomValue::EcdsaSecp256k1Signature(value) => serialize_type_name_tagged_value(
+            serializer,
+            context,
+            "EcdsaSecp256k1Signature",
+            &format!("{}", value),
+        ),
+        ScryptoCustomValue::EddsaEd25519PublicKey(value) => serialize_type_name_tagged_value(
+            serializer,
+            context,
+            "EddsaEd25519PublicKey",
+            &format!("{}", value),
+        ),
+        ScryptoCustomValue::EddsaEd25519Signature(value) => serialize_type_name_tagged_value(
+            serializer,
+            context,
+            "EddsaEd25519Signature",
+            &format!("{}", value),
+        ),
+        ScryptoCustomValue::Decimal(value) => {
+            // The fact it's a decimal number will be obvious from context, so favour simplicity over verbosity
+            serialize_potentially_transparent_value(
+                serializer,
+                context,
+                "Decimal",
+                &format!("{}", value),
+            )
+        }
+        ScryptoCustomValue::PreciseDecimal(value) => {
+            // The fact it's a decimal number will be obvious from context, so favour simplicity over verbosity
+            serialize_potentially_transparent_value(
+                serializer,
+                context,
+                "PreciseDecimal",
+                &format!("{}", value),
+            )
+        }
+        ScryptoCustomValue::NonFungibleId(value) => serialize_type_name_tagged_value(
+            serializer,
+            context,
+            "NonFungibleId",
+            &format!("{}", value),
+        ),
+    }
+}
+
+pub fn serialize_potentially_transparent_value<S: Serializer, T: Serialize>(
+    serializer: S,
+    context: &ScryptoValueFormattingContext,
+    type_name: &'static str,
+    value: &T,
+) -> Result<S::Ok, S::Error> {
+    match context.serialization_type {
+        ScryptoValueSerializationType::Simple => value.serialize(serializer),
+        ScryptoValueSerializationType::Invertible => {
+            // In the invertible case, we have to add the tag anyway
+            serialize_type_name_tagged_value(serializer, context, type_name, value)
+        }
+    }
+}
+
+pub fn serialize_type_name_tagged_value<S: Serializer, T: Serialize + ?Sized>(
+    serializer: S,
+    _context: &ScryptoValueFormattingContext,
+    type_name: &'static str,
```

Comment:
Possibly to use `ScryptoTypeId`?

Reply:
I considered it, but it's possibly less performant - happy to do it if you'd rather though... will do it in the next PR :)

Reply:
I suppose Rust simple enum matching is performant. Would prefer enum for compiler check to avoid mistakes, considering we're constantly renaming things. :)

Reply:
I've changed this now 👍